### PR TITLE
[GPU] optimized 5d linear_onnx and nearest mode and tests

### PR DESCRIPTION
**Pull request by OlehKravchyshyn**
Performance optimization for 5d ONNX linear blocked formats for f32/16 and u8/i8 types:
       - adjusting interpolate/resample GPU implementation for supporting u8/i8
       - adjusting resample opt kernels implementations to support new features
       - wrote performance tests as unit tests

### CODE REVIEW

**Comment by OlehKravchyshyn** [comment]
Hi @wilson-seok 
this pull request is ready for performance tests
Locally we get such results listed below running the command
clDNN_unit_tests64 --gtest_filter=*resample_opt_perf_linear_5_onnx* 

Results:
Exectued time  optimized  input([b:1, f:32, x:64, y:64, z:5, w:1, g:1]) output([b:1, f:32, x:128, y:128, z:5, w:1, g:1]) b_fs_zyx_fsv16 f16 1.987
Exectued time  not optimazed  input([b:1, f:32, x:64, y:64, z:5, w:1, g:1]) output([b:1, f:32, x:128, y:128, z:5, w:1, g:1]) bfzyx f16 10.4712

Exectued time  optimized  input([b:1, f:32, x:64, y:64, z:5, w:1, g:1]) output([b:1, f:32, x:128, y:128, z:5, w:1, g:1]) b_fs_zyx_fsv16 f32 1.834
Exectued time  not optimazed  input([b:1, f:32, x:64, y:64, z:5, w:1, g:1]) output([b:1, f:32, x:128, y:128, z:5, w:1, g:1]) bfzyx f32 10.5286

Exectued time  optimized  input([b:1, f:32, x:64, y:64, z:5, w:1, g:1]) output([b:1, f:32, x:128, y:128, z:5, w:1, g:1]) bs_fs_zyx_bsv16_fsv16 f16 1.9848
Exectued time  not optimazed  input([b:1, f:32, x:64, y:64, z:5, w:1, g:1]) output([b:1, f:32, x:128, y:128, z:5, w:1, g:1]) bfzyx f16 10.8441

Exectued time  optimized  input([b:1, f:32, x:64, y:64, z:5, w:1, g:1]) output([b:1, f:32, x:128, y:128, z:5, w:1, g:1]) bs_fs_zyx_bsv16_fsv16 f32 2.3429
Exectued time  not optimazed  input([b:1, f:32, x:64, y:64, z:5, w:1, g:1]) output([b:1, f:32, x:128, y:128, z:5, w:1, g:1]) bfzyx f32 10.5617

Exectued time  optimized  input([b:1, f:32, x:64, y:64, z:5, w:1, g:1]) output([b:1, f:32, x:128, y:128, z:5, w:1, g:1]) bs_fs_zyx_bsv32_fsv16 f16 1.7191
Exectued time  not optimazed  input([b:1, f:32, x:64, y:64, z:5, w:1, g:1]) output([b:1, f:32, x:128, y:128, z:5, w:1, g:1]) bfzyx f16 9.3731

Exectued time  optimized  input([b:1, f:32, x:64, y:64, z:5, w:1, g:1]) output([b:1, f:32, x:128, y:128, z:5, w:1, g:1]) bs_fs_zyx_bsv32_fsv16 f32 2.2083
Exectued time  not optimazed  input([b:1, f:32, x:64, y:64, z:5, w:1, g:1]) output([b:1, f:32, x:128, y:128, z:5, w:1, g:1]) bfzyx f32 10.631

Exectued time  optimized  input([b:1, f:32, x:64, y:64, z:5, w:1, g:1]) output([b:1, f:32, x:128, y:128, z:5, w:1, g:1]) b_fs_zyx_fsv32 i8 2.5521
Exectued time  not optimazed  input([b:1, f:32, x:64, y:64, z:5, w:1, g:1]) output([b:1, f:32, x:128, y:128, z:5, w:1, g:1]) bfzyx i8 9.2773

Exectued time  optimized  input([b:1, f:32, x:64, y:64, z:5, w:1, g:1]) output([b:1, f:32, x:128, y:128, z:5, w:1, g:1]) b_fs_zyx_fsv32 u8 3.0447
Exectued time  not optimazed  input([b:1, f:32, x:64, y:64, z:5, w:1, g:1]) output([b:1, f:32, x:128, y:128, z:5, w:1, g:1]) bfzyx u8 5.9453

Exectued time  optimized  input([b:1, f:32, x:64, y:64, z:5, w:1, g:1]) output([b:1, f:32, x:128, y:128, z:5, w:1, g:1]) bs_fs_zyx_bsv16_fsv32 i8 1.852
Exectued time  not optimazed  input([b:1, f:32, x:64, y:64, z:5, w:1, g:1]) output([b:1, f:32, x:128, y:128, z:5, w:1, g:1]) bfzyx i8 8.5692

Exectued time  optimized  input([b:1, f:32, x:64, y:64, z:5, w:1, g:1]) output([b:1, f:32, x:128, y:128, z:5, w:1, g:1]) bs_fs_zyx_bsv16_fsv32 u8 1.703
Exectued time  not optimazed  input([b:1, f:32, x:64, y:64, z:5, w:1, g:1]) output([b:1, f:32, x:128, y:128, z:5, w:1, g:1]) bfzyx u8 6.2551

Exectued time  optimized  input([b:1, f:32, x:64, y:64, z:5, w:1, g:1]) output([b:1, f:32, x:128, y:128, z:5, w:1, g:1]) bs_fs_zyx_bsv32_fsv32 i8 1.9067
Exectued time  not optimazed  input([b:1, f:32, x:64, y:64, z:5, w:1, g:1]) output([b:1, f:32, x:128, y:128, z:5, w:1, g:1]) bfzyx i8 8.6603

Exectued time  optimized  input([b:1, f:32, x:64, y:64, z:5, w:1, g:1]) output([b:1, f:32, x:128, y:128, z:5, w:1, g:1]) bs_fs_zyx_bsv32_fsv32 u8 1.9214
Exectued time  not optimazed  input([b:1, f:32, x:64, y:64, z:5, w:1, g:1]) output([b:1, f:32, x:128, y:128, z:5, w:1, g:1]) bfzyx u8 6.4954

File: src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/resample/resample_kernel_opt.cpp [hunk 1]
```diff
@@ -116,17 +139,33 @@ bool ResampleKernelOpt::Validate(const Params& p, const optional_params& o) cons
...
-        input.GetLayout() != DataLayout::bs_fs_yx_bsv32_fsv32)
+        input.GetLayout() != DataLayout::bs_fs_yx_bsv32_fsv32 &&
+
+        input.GetLayout() != DataLayout::b_fs_zyx_fsv16 &&
+        input.GetLayout() != DataLayout::bs_fs_zyx_bsv16_fsv16 &&
+        input.GetLayout() != DataLayout::bs_fs_zyx_bsv32_fsv16 &&
+        input.GetLayout() != DataLayout::b_fs_zyx_fsv32 &&
+        input.GetLayout() != DataLayout::bs_fs_zyx_bsv16_fsv32 &&
+        input.GetLayout() != DataLayout::bs_fs_zyx_bsv32_fsv32)
+        return false;
```
Comment (tgubanova-lohika) [inline src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/resample/resample_kernel_opt.cpp:162]: i think this if is not needed at all as it duplicates
EnableInputLayout logic

**Review by tgubanova-lohika — COMMENTED** [review]
better to rename commit name I think (no critical after merge but little bit confusing now)

**Comment by wilson-seok** [comment]
@OlehKravchyshyn As you commented, this PR is including commit of #12821 . Then will you update this PR after merging of it? Or review and merge this PR only? We are leaving some review comments on #12821.

**Comment by tgubanova-lohika** [comment]
> @OlehKravchyshyn As you commented, this PR is including commit of #12821 . Then will you update this PR after merging of it? Or review and merge this PR only? We are leaving some review comments on #12821.

I believe it's  better to merge #12821  as formally there are different tasks, so better to have separate merge for reference 5D support firstly. After the merge, this PR will be rebased (the same would be for nearest 5d optimization)

File: src/plugins/intel_gpu/tests/test_cases/resample_gpu_test.cpp [hunk 2]
```diff
@@ -1942,10 +2224,30 @@ INSTANTIATE_TEST_SUITE_P(resample_opt_smoke_linear_onnx,
...
                                 { data_types::f16, {1, 128, 13, 13},  {1, 128, 26, 26},  1, resample::InterpolateOp::InterpolateMode::LINEAR_ONNX, 1, format::bs_fs_yx_bsv32_fsv32, format::bs_fs_yx_bsv32_fsv32, {}, {}},
                                 { data_types::f16, {1, 128, 13, 13},  {1, 128, 26, 26},  1, resample::InterpolateOp::InterpolateMode::LINEAR_ONNX, 1, format::b_fs_yx_fsv16, format::b_fs_yx_fsv32, {}, {}},
                             }
                         ));
+
+INSTANTIATE_TEST_SUITE_P(resample_opt_perf_linear_5_onnx,
+                         resample_opt_random_test_ext,
+                         testing::ValuesIn(
+                            std::vector<resample_opt_random_test_params>{
+                                { data_types::f16, {1, 32, 64, 64, 5},  {1, 32, 128, 128, 5},  1, resample::InterpolateOp::InterpolateMode::LINEAR_ONNX, 1, format::b_fs_zyx_fsv16, format::b_fs_zyx_fsv16, {}, {}},
```
Comment (jade-cho) [inline src/plugins/intel_gpu/tests/test_cases/resample_gpu_test.cpp:2239]: I've tested this PR. Does it work as you intended? It looks like you wanted to compare opt kernel with ref kernel. But bfzyx format is selected for ref kernel. Please test with OV_GPU_DumpGraphs test variable. Please check this out again.

File: src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/resample/resample_kernel_opt.cpp (hunk 1 above)
Comment (OlehKravchyshyn) [reply src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/resample/resample_kernel_opt.cpp:162]: yes, you right

File: src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/resample_ref.cl [hunk 3]
```diff
@@ -286,21 +290,19 @@ KERNEL (resample_gpu_ref)(__global INPUT0_TYPE* input,
     bool trOutOfBounds = in_y1 < 0 || in_y1 >= in_size[3] || in_x2 < 0 || in_x2 >= in_size[4];
     bool blOutOfBounds = in_y2 < 0 || in_y2 >= in_size[3] || in_x1 < 0 || in_x1 >= in_size[4];
     bool brOutOfBounds = in_y2 < 0 || in_y2 >= in_size[3] || in_x2 < 0 || in_x2 >= in_size[4];
-#endif
+
```
[the remark this answers is not in the document]
Comment (OlehKravchyshyn) [reply src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/resample_ref.cl:293]: copied from PR #12821 
Seems it was wrong to get a value from input and then check if it is out of bounds.
If there is padding then in_f, in_y1, in_x1 coordinates may be negative or be far from valid coordinates, so INPUT0_GET_INDEX(batch, in_f, in_y1, in_x1) may return a negative index or too big (experienced than while debugged 5d formats).

File: src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/resample_ref.cl [hunk 4]
```diff
@@ -318,6 +320,108 @@ KERNEL (resample_gpu_ref)(__global INPUT0_TYPE* input,
...
+    OUTPUT_TYPE x211 = input[INPUT0_GET_INDEX(batch, feature, in_z1, in_y1, in_x2)];
+    OUTPUT_TYPE x121 = input[INPUT0_GET_INDEX(batch, feature, in_z1, in_y2, in_x1)];
+    OUTPUT_TYPE x221 = input[INPUT0_GET_INDEX(batch, feature, in_z1, in_y2, in_x2)];
+    OUTPUT_TYPE x112 = input[INPUT0_GET_INDEX(batch, feature, in_z2, in_y1, in_x1)];
+    OUTPUT_TYPE x212 = input[INPUT0_GET_INDEX(batch, feature, in_z2, in_y1, in_x2)];
+    OUTPUT_TYPE x122 = input[INPUT0_GET_INDEX(batch, feature, in_z2, in_y2, in_x1)];
+    OUTPUT_TYPE x222 = input[INPUT0_GET_INDEX(batch, feature, in_z2, in_y2, in_x2)];
+#endif
+
+//    printf("oz(%d) oy(%d) ox(%d) dx1(%f) dx2(%f) dy1(%f) dy2(%f) dz1(%f) dz2(%f)\n", oz, oy, ox, dx1, dx2, dy1, dy2, dz1, dz2);
```
[the remark this answers is not in the document]
Comment (OlehKravchyshyn) [reply src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/resample_ref.cl]: done

File: src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/resample_ref.cl [hunk 5]
```diff
@@ -318,6 +320,108 @@ KERNEL (resample_gpu_ref)(__global INPUT0_TYPE* input,
 #endif
         output[OUTPUT_GET_INDEX(batch, in_f, oy, ox)] = res;
     }
+#endif
```
[the remark this answers is not in the document]
Comment (OlehKravchyshyn) [reply src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/resample_ref.cl]: done

File: src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/resample_ref.cl [hunk 6]
```diff
@@ -46,6 +46,8 @@ inline int FUNC(get_nearest_val)(float num, bool is_downsample)
 
 inline float FUNC(get_original_coordinate)(float num, float scale, int length_resized, int length_original)
 {
+    if (scale == 1.0f)
```
[the remark this answers is not in the document]
Comment (OlehKravchyshyn) [reply src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/resample_ref.cl:49]: copied from PR https://github.com/openvinotoolkit/openvino/pull/12821
scales are always calculated in ResampleKernelBase::GetJitConstants, for scaled and sized shape calc mode, looks the scale value are always valid.
This change was necessary for passing tf_half_pixel_for_nn transform mode tests.
For 4d formats, those tests are disabled in skip_tests_config.cpp file (look for .*Interpolate.*cubic.*tf_half_pixel_for_nn.FP16. line).
The reference implementation of calculating the original coordinate also checks the scale before further calculation
Look at
class GetOriginalCoordinate final {
...
float operator()(float x_resized, float x_scale, float length_resized, float length_original) const {
if (x_scale == 1.0f || (length_resized == length_original)) {
return x_resized;
}
return m_func(x_resized, x_scale, length_resized, length_original);
}
...
};

File: src/plugins/intel_gpu/src/plugin/ops/interpolate.cpp [hunk 7]
```diff
@@ -50,18 +50,23 @@ static void CreateInterpolateOp(Program& p, const std::shared_ptr<ngraph::op::v4
     if (axes.size() != scales.size())
         IE_THROW() << op->get_friendly_name() << " Incorrect axes and scales should be the same size";
 
+    // TODO shouldn't be all this checking done in ngraph::op::v4::Interpolate?
```
[the remark this answers is not in the document]
Comment (OlehKravchyshyn) [reply src/plugins/intel_gpu/src/plugin/ops/interpolate.cpp:53]: removed comment

File: src/plugins/intel_gpu/tests/test_cases/resample_gpu_test.cpp (hunk 2 above)
Comment (OlehKravchyshyn) [reply src/plugins/intel_gpu/tests/test_cases/resample_gpu_test.cpp:2239]: yes, that is right. We discussed this during the last meeting. The current commit contains three tests for each format:
- opt for blocked format
- ref for blocked format
- ref for planar format

**Comment by wilson-seok** [comment]
@OlehKravchyshyn Please rebase the PR branch to the latest one. Current build error about wrong repo(404 error) requires recent commit in OpenVINO master. Thanks!

**Comment by OlehKravchyshyn** [comment]
> @OlehKravchyshyn Please rebase the PR branch to the latest one. Current build error about wrong repo(404 error) requires recent commit in OpenVINO master. Thanks!
rebased

File: src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/resample/resample_kernel_opt.cpp [hunk 8]
```diff
@@ -116,17 +139,19 @@ bool ResampleKernelOpt::Validate(const Params& p, const optional_params& o) cons
...
-        input.GetLayout() != DataLayout::b_fs_yx_fsv16 &&
-        input.GetLayout() != DataLayout::b_fs_yx_fsv32 &&
-        input.GetLayout() != DataLayout::bs_fs_yx_bsv32_fsv16 &&
-        input.GetLayout() != DataLayout::bs_fs_yx_bsv32_fsv32)
+    if (params.resampleType == ResampleType::LINEAR_ONNX &&
+        input.Dimentions() == 5 &&
+        (input.Batch().v != output.Batch().v ||
+        input.Feature().v != output.Feature().v ||
+        input.Z().v != output.Z().v))
         return false;
```
Comment (jade-cho) [inline src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/resample/resample_kernel_opt.cpp:156]: This PR only supports nearest resample type for 5D. Right? In the case of the opt kernel, 5D tensors were not previously supported. It seems that the Validate() function for other resample types needs to return false if input/output tensor is 5D tensor.

File: src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/resample/resample_kernel_opt.cpp [hunk 9]
```diff
@@ -116,17 +139,19 @@ bool ResampleKernelOpt::Validate(const Params& p, const optional_params& o) cons
...
-        params.resampleType != ResampleType::BILINEAR_INTERP)
+        params.resampleType != ResampleType::BILINEAR_INTERP &&
+        params.resampleType != ResampleType::LINEAR_ONNX)
         return false;
 
-    if (input.GetLayout() != DataLayout::fs_b_yx_fsv32 &&
-        input.GetLayout() != DataLayout::b_fs_yx_fsv16 &&
-        input.GetLayout() != DataLayout::b_fs_yx_fsv32 &&
-        input.GetLayout() != DataLayout::bs_fs_yx_bsv32_fsv16 &&
-        input.GetLayout() != DataLayout::bs_fs_yx_bsv32_fsv32)
```
Comment (jade-cho) [inline src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/resample/resample_kernel_opt.cpp:129]: Does it mean the size of axis z should be same between input and output when input is 5D? According to the specification, when it is a 5D tensor, it is possible to resample the x, y, and z axes.

File: src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/resample/resample_kernel_opt.cpp (hunk 8 above)
Comment (OlehKravchyshyn) [reply src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/resample/resample_kernel_opt.cpp:158]: Added checking of nearest resample for this checking

File: src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/resample/resample_kernel_opt.cpp (hunk 9 above)
Comment (OlehKravchyshyn) [reply src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/resample/resample_kernel_opt.cpp:129]: We got a list of shapes that must be optimized for linear_onnx, so we added support only for resampling X and Y axes.
Resampling Z axes can be added to the existing resampl_opt kernel but will decrease the existing performance for requested shapes.

During today's status meeting, we agreed to add a resampling Z axe as separate PR, for this purpose will be created new CPP and cl kernel files.

**Review by wilson-seok — APPROVED** [review]

File: src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/resample/resample_kernel_base.cpp [hunk 10]
```diff
@@ -231,6 +231,7 @@ KernelsData ResampleKernelBase::GetCommonKernelsData(const Params& params, const
     FillCLKernelData(kernel, dispatchData, params.engineInfo, kernelName, jit, entry_point,
                      DEFAULT, false, false, 1, GetFusedPrimitiveInputsCount(params));
 
+    //std::cout << kernel.code.kernelString->get_str() << std::endl;
```
[the remark this answers is not in the document]
Comment (OlehKravchyshyn) [reply src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/resample/resample_kernel_base.cpp:234]: fixed, it was leftover debugging, it didn't work after rebasing

**Review by jade-cho — APPROVED** [review]

**Comment by wilson-seok** [comment]
@OlehKravchyshyn Fix for openvino-lin-arm64 error is now merged. Please rebase this PR to pass it.

**Comment by wilson-seok** [comment]
@OlehKravchyshyn Please resolve the conflicts.

**Comment by OlehKravchyshyn** [comment]
Seems checking format types in ResampleKernelOpt::Validate, e.g.
if (arg.outputs[0].GetLayout() == DataLayout::bs_fs_yx_bsv16_fsv16 ||
...
            arg.outputs[0].GetLayout() == DataLayout::bs_fs_zyx_bsv32_fsv32) {
    return false
}
duplicate logic from GetSupportedKey()
That is why it was removed from

**Comment by jade-cho** [comment]
Please resolve the conflicts.

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/resample_opt.cl [hunk 11]
```diff
@@ -251,9 +261,15 @@ KERNEL (resample_opt)(__global INPUT0_TYPE* input,
...
 
+#if OUTPUT_DIMS == 5
+        const int iz = floor(z * SCALES[2]);
+        in_vec_t res = READ_FUNC(input, INPUT0_GET_INDEX(b, feature_block, iz, iy, ix));
+#else
         in_vec_t res = READ_FUNC(input, INPUT0_GET_INDEX(b, feature_block, iy, ix));
+#endif
 #elif defined(SAMPLE_TYPE_INTERP)
-    unroll_for (uint out_x = 0; out_x < OUTPUT_X_BLOCK_SIZE; out_x++) {
+    unroll_for (uint out_x = 0; out_x#if defined(SAMPLE_TYPE_CAFFE_INTERP)
```
Comment (jade-cho) [inline src/plugins/intel_gpu/src/kernel_selector/cl_kernels/resample_opt.cl:271]: What is this line?

**Review by jade-cho — CHANGES_REQUESTED** [review]

**Comment by jade-cho** [comment]
@OlehKravchyshyn Are they separated into another PR for different resample types for 5D tensor? If it has did, please let me know the PR.

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/resample_opt.cl (hunk 11 above)
Comment (OlehKravchyshyn) [reply src/plugins/intel_gpu/src/kernel_selector/cl_kernels/resample_opt.cl:271]: something happened by accident, strange is that tests didn't fail. Fixed

**Review by jade-cho — APPROVED** [review]

**Comment by jade-cho** [comment]
@OlehKravchyshyn Please resolve the conflicts.

**Pull request merged**